File: backend/src/modules/twitter-api.ts
import { ApiError, buildErrorFromResponse, writeErrorToResponse } from '../models/errors'
import { DocumentsDatabase } from '../db/documents'
import { HttpConfig } from './http'
import { NewPostRequest, NewPostResponse, UnvalidatedNewPostRequest } from '../models/posts'
import { Request, Response } from 'express'
import crypto from 'crypto'
import logger from '../utils/logger'
import OAuth from 'oauth-1.0a'
import qs from 'querystring'
import result, { Result } from '../models/result'
import utils from '../utils/text'
import { FilesModule } from './files'

export type TwitterAuthConfig = {
  twitterOauth1ConsumerKey: string
  twitterOauth1ConsumerSecret: string
}

type CreateNewPostRequest = {
  text: string
  media?: { media_ids: string[] }
}

type TwitterMediaUploadResponse = {
  id: string
}

type AuthorizedToken = {
  key: string
  secret: string
}

const authorizeURL = new URL('https://api.twitter.com/oauth/authorize')
const accessTokenURL = 'https://api.twitter.com/oauth/access_token'

const callback = (baseUrl: string, jwtAccessToken: string) =>
  `${baseUrl}/api/twitter/callback?access_token=${encodeURIComponent(jwtAccessToken)}`
const requestTokenURL = (baseUrl: string, jwtAccessToken: string) =>
  `https://api.twitter.com/oauth/request_token?oauth_callback=${encodeURIComponent(callback(baseUrl, jwtAccessToken))}&x_auth_access_type=write`

export class TwitterApiModule {
  constructor(
    private config: TwitterAuthConfig & HttpConfig,
    private docsDb: DocumentsDatabase,
    private files: FilesModule
  ) {}

  private oauth = new OAuth({
    consumer: {
      key: this.config.twitterOauth1ConsumerKey,
      secret: this.config.twitterOauth1ConsumerSecret
    },
    signature_method: 'HMAC-SHA1',
    hash_function: (baseString, key) =>
      crypto.createHmac('sha1', key).update(baseString).digest('base64')
  })

  private requestToken = async (jwtAccessToken: string): Promise<Result<any, ApiError>> => {
    const tokenUrl = requestTokenURL(this.config.baseUrl, jwtAccessToken)
    const authHeader = this.oauth.toHeader(
      this.oauth.authorize({
        url: tokenUrl,
        method: 'POST'
      })
    )
    const response = await fetch(tokenUrl, {
      method: 'POST',
      headers: {
        Authorization: authHeader['Authorization']
      }
    })
    if (!response.ok) {
      return result.error(await buildErrorFromResponse(response, 'twitter'))
    }
    const body = await response.text()
    const parsed = qs.parse(body)
    return result.success({
      oauth_token: parsed.oauth_token,
      oauth_token_secret: parsed.oauth_token_secret
    })
  }

  buildAuthorizeURL = async (jwtAccessToken: string): Promise<Result<string, ApiError>> => {
    const requestToken = await this.requestToken(jwtAccessToken)
    if (requestToken.type === 'error') {
      return requestToken
    }
    authorizeURL.searchParams.set('oauth_token', requestToken.result.oauth_token)
    return result.success(authorizeURL.href)
  }

  private buildAuthHeader = (url: string, token: AuthorizedToken): OAuth.Header =>
    this.oauth.toHeader(
      this.oauth.authorize(
        {
          url,
          method: 'POST'
        },
        token
      )
    )

  private restoreOauthTokenFromDb = async () => {
    const doc = await this.docsDb.searchByKey('twitter-oauth-token')
    if (doc && doc.payload) {
      return JSON.parse(doc.payload) as AuthorizedToken
    }
    return null
  }

  hasTwitterAuth = async (): Promise<boolean> => {
    logger.info('Checking Twitter auth status...')
    const token = await this.restoreOauthTokenFromDb()
    logger.info('Token taken from DB:', token)
    return !!token
  }

  saveOauthToken = async (token: string, verifier: string): Promise<Result<void, ApiError>> => {
    const authHeader = this.oauth.toHeader(
      this.oauth.authorize({
        url: accessTokenURL,
        method: 'POST'
      })
    )
    const path = `https://api.twitter.com/oauth/access_token?oauth_verifier=${verifier}&oauth_token=${token}`
    const req = await fetch(path, {
      method: 'GET',
      headers: {
        Authorization: authHeader['Authorization']
      }
    })
    if (!req.ok) {
      return result.error(
        await buildErrorFromResponse(req, 'twitter', {
          action: 'save-oauth'
        })
      )
    }
    const body = qs.parse(await req.text())
    const authorizedToken: AuthorizedToken = {
      key: body.oauth_token as string,
      secret: body.oauth_token_secret as string
    }
    await this.docsDb.createOrUpdate(
      'twitter-oauth-token',
      JSON.stringify(authorizedToken),
      'twitter-oauth-token'
    )
    return result.success(undefined)
  }

  authorizeHttpRoute = async (req: Request, res: Response) => {
    if (!req.jwtToken) {
      return res.status(401).send({ error: 'Unauthorized' })
    }
    const url = await this.buildAuthorizeURL(req.jwtToken)
    if (url.type === 'error') return writeErrorToResponse(res, url.error)
    res.redirect(url.result)
  }

  authCallbackHttpRoute = async (req: Request, res: Response) => {
    const token = req.query.oauth_token as string | undefined
    const verifier = req.query.oauth_verifier as string | undefined
    if (!token || !verifier) {
      return res.status(400).send({ error: 'Invalid request' })
    }

    logger.info('Twitter auth callback:', { token, verifier })
    const r = await this.saveOauthToken(token, verifier)
    if (r.type === 'error') return writeErrorToResponse(res, r.error)
    res
      .set('Cache-Control', 'no-store, no-cache, must-revalidate, private')
      .set('Pragma', 'no-cache')
      .set('Expires', '0')
      .redirect('/account')
  }

  statusHttpRoute = async (_req: Request, res: Response) => {
    const row = await this.docsDb.searchByKey('twitter-oauth-token')
    if (!row) return res.status(404).send({ error: 'Not found' })
    return res.status(200).send({
      createdAt: row.createdAt.getTime()
    })
  }

  uploadMedia =
    (token: AuthorizedToken) =>
    async (uuid: string): Promise<Result<TwitterMediaUploadResponse, ApiError>> => {
      try {
        const r = await this.files.readImageFile(uuid)
        if (r === null)
          return result.error({
            type: 'validation-error',
            module: 'twitter',
            status: 404,
            error: 'Failed to read image file — uuid: ' + uuid
          })

        const url = 'https://upload.twitter.com/1.1/media/upload.json'
        const authHeader = this.buildAuthHeader(url, token)
        const formData = new FormData()
        formData.append('media', new Blob([r.bytes], { type: r.mimetype }), r.originalname)
        formData.append('media_category', 'tweet_image')
        const response = await fetch(url, {
          method: 'POST',
          headers: {
            Authorization: authHeader['Authorization']
          },
          body: formData
        })
        if (!response.ok) {
          return result.error(
            await buildErrorFromResponse(response, 'twitter', {
              action: 'upload-media'
            })
          )
        }
        const data = await response.json()
        const mediaIdString = data['media_id_string'] as string

        if (r.altText) {
          const altTextUrl = 'https://api.twitter.com/1.1/media/metadata/create.json'
          const authHeader = this.buildAuthHeader(altTextUrl, token)
          const altTextResponse = await fetch(altTextUrl, {
            method: 'POST',
            headers: {
              Authorization: authHeader['Authorization'],
              'Content-Type': 'application/json'
            },
            body: JSON.stringify({
              media_id: mediaIdString,
              alt_text: { text: r.altText }
            })
          })
          if (!altTextResponse.ok) {
            return result.error(
              await buildErrorFromResponse(altTextResponse, 'twitter', {
                action: 'upload-media-alt-text'
              })
            )
          }
        }

        return result.success({
          id: mediaIdString
        })
      } catch (err) {
        logger.error(`Failed to upload media (twitter) — uuid ${uuid}:`, err)
        return result.error({
          type: 'caught-exception',
          error: `Failed to upload media — uuid: ${uuid}`,
          status: 500,
          module: 'twitter'
        })
      }
    }

  createPost = async (post: NewPostRequest): Promise<Result<NewPostResponse, ApiError>> => {
    const createPostURL = 'https://api.twitter.com/2/tweets'
    const token = await this.restoreOauthTokenFromDb()
    if (!token) {
      return result.error({
        type: 'validation-error',
        status: 401,
        error: 'Unauthorized: Missing Twitter OAuth token!',
        module: 'twitter'
      })
    }

    const images: string[] = []
    if (post.images) {
      const rs = await Promise.all(post.images.map(this.uploadMedia(token)))
      logger.info('Uploaded images:', rs)

      for (const r of rs) {
        if (r.type === 'success') {
          images.push(r.result.id)
        } else {
          return result.error({
            type: 'composite-error',
            status: 502,
            module: 'twitter',
            error: 'Failed to upload images.',
            responses: rs
          })
        }
      }
    }

    const status =
      (post.cleanupHtml ? utils.convertHtml(post.content) : post.content.trim()) +
      (post.link ? `\n\n${post.link}` : '')

    // Create the post
    const data: CreateNewPostRequest = {
      text: status
    }
    if (images.length > 0) {
      data.media = { media_ids: images }
    }
    const authHeader = this.buildAuthHeader(createPostURL, token)
    const response = await fetch(createPostURL, {
      method: 'POST',
      headers: {
        Authorization: authHeader['Authorization'],
        'Content-Type': 'application/json',
        Accept: 'application/json'
      },
      body: JSON.stringify(data)
    })
    if (!response.ok) {
      return result.error(
        await buildErrorFromResponse(response, 'twitter', {
          action: 'create-post'
        })
      )
    }

    const body = await response.json()
    logger.info('Twitter create-post response:', body)
    return result.success({
      module: 'twitter',
      id: body.data.id
    })
  }

  createPostRoute = async (
    post: UnvalidatedNewPostRequest
  ): Promise<Result<NewPostResponse, ApiError>> => {
    const content = post.content
    if (!content) {
      return result.error({
        type: 'validation-error',
        status: 400,
        error: 'Bad Request: Missing content!',
        module: 'twitter'
      })
    }

    return await this.createPost({ ...post, content })
  }

  createPostHttpRoute = async (req: Request, res: Response) => {
    const r = await this.createPostRoute(req.body)
    if (r.type === 'success') {
      res.status(200).send(r.result)
    } else {
      writeErrorToResponse(res, r.error)
    }
  }
}

File: backend-scala/src/main/scala/socialpublish/api/TwitterApi.scala
package socialpublish.api

import cats.effect.*
import cats.syntax.all.*
import io.circe.*
import io.circe.syntax.*
import io.circe.parser.*
import org.http4s.*
import org.http4s.client.Client
import org.http4s.circe.CirceEntityEncoder.*
import org.http4s.circe.CirceEntityDecoder.*
import org.http4s.multipart.{Multipart, Part}
import org.http4s.headers.`Content-Type`
import org.typelevel.ci.CIStringSyntax
import org.typelevel.log4cats.Logger
import fs2.Stream
import socialpublish.config.AppConfig
import socialpublish.db.DocumentsDatabase
import socialpublish.models.*
import socialpublish.services.FilesService
import socialpublish.utils.TextUtils

import java.util.UUID
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import java.util.Base64
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import scala.annotation.nowarn
import scala.annotation.unused
import scala.util.Random

// Twitter API implementation with OAuth 1.0a
trait TwitterApi {
  def createPost(request: NewPostRequest): Result[NewPostResponse]
  def getAuthorizationUrl(accessToken: String): Result[String]
  def handleCallback(oauthToken: String, oauthVerifier: String): Result[Unit]
  def hasTwitterAuth: IO[Boolean]
  def getAuthStatus: IO[Option[Long]]
}

object TwitterApi {
  def apply(
      config: AppConfig,
      client: Client[IO],
      files: FilesService,
      docsDb: DocumentsDatabase,
      logger: Logger[IO]
  ): TwitterApi =
    new TwitterApiImpl(config, client, files, docsDb, logger)
}

private case class AuthorizedToken(
    key: String,
    secret: String
) derives Codec.AsObject
private case class TwitterMediaUploadResponse(
    media_id_string: String
) derives Codec.AsObject
private case class CreateNewPostRequest(
    text: String,
    media: Option[MediaIds]
) derives Codec.AsObject
private case class MediaIds(
    media_ids: List[String]
) derives Codec.AsObject
private case class TweetData(
    id: String
) derives Codec.AsObject
private case class TweetResponse(
    data: TweetData
) derives Codec.AsObject

private class TwitterApiImpl(
    config: AppConfig,
    client: Client[IO],
    files: FilesService,
    docsDb: DocumentsDatabase,
    logger: Logger[IO]
) extends TwitterApi {

  private val requestTokenURL = "https://api.twitter.com/oauth/request_token"
  private val authorizeURL = "https://api.twitter.com/oauth/authorize"
  private val accessTokenURL = "https://api.twitter.com/oauth/access_token"
  private val createTweetURL = "https://api.twitter.com/2/tweets"
  private val mediaUploadURL = "https://upload.twitter.com/1.1/media/upload.json"
  private val altTextURL = "https://api.twitter.com/1.1/media/metadata/create.json"

  override def hasTwitterAuth: IO[Boolean] =
    restoreOauthTokenFromDb.map(_.isDefined)

  override def getAuthStatus: IO[Option[Long]] =
    docsDb.searchByKey("twitter-oauth-token").map(_.map(_.createdAt.getEpochSecond * 1000))

  override def getAuthorizationUrl(jwtAccessToken: String): Result[String] = {
    val callbackUrl =
      s"${config.baseUrl}/api/twitter/callback?access_token=${URLEncoder.encode(jwtAccessToken, "UTF-8")}"
    val reqUrl =
      s"$requestTokenURL?oauth_callback=${URLEncoder.encode(callbackUrl, "UTF-8")}&x_auth_access_type=write"
    val oauthParams = generateOAuthParams("POST", reqUrl, Map.empty, None)
    val authHeader = buildAuthorizationHeader(oauthParams)

    val request = Request[IO](Method.POST, Uri.unsafeFromString(reqUrl))
      .withHeaders(Header.Raw(ci"Authorization", authHeader))

    for {
      response <- Result.liftIO(
        client.expect[String](request).flatMap { body =>
          IO {
            val params = parseQueryString(body)
            params.get("oauth_token") match {
              case Some(token) => token
              case None => throw new Exception("No oauth_token in response")
            }
          }
        }
      )

      authUrl = s"$authorizeURL?oauth_token=$response"
    } yield authUrl
  }

  override def handleCallback(oauthToken: String, oauthVerifier: String): Result[Unit] = {
    val oauthParams = generateOAuthParams("POST", accessTokenURL, Map.empty, None)
    val authHeader = buildAuthorizationHeader(oauthParams)
    val url = s"$accessTokenURL?oauth_verifier=$oauthVerifier&oauth_token=$oauthToken"
    val request = Request[IO](Method.GET, Uri.unsafeFromString(url))
      .withHeaders(Header.Raw(ci"Authorization", authHeader))

    Result.liftIO(
      client.expect[String](request).flatMap { body =>
        val params = parseQueryString(body)
        val token = AuthorizedToken(
          key = params.getOrElse("oauth_token", ""),
          secret = params.getOrElse("oauth_token_secret", "")
        )

        docsDb.createOrUpdate(
          kind = "twitter-oauth-token",
          payload = token.asJson.noSpaces,
          tags = List(DocumentTag("twitter-oauth-token", "key"))
        ).void
      }
    )
  }

  override def createPost(postReq: NewPostRequest): Result[NewPostResponse] =
    for {
      token <- Result.liftIO(restoreOauthTokenFromDb).flatMap {
        case Some(t) => Result.success(t)
        case None => Result.error(ApiError.unauthorized("Missing Twitter OAuth token"))
      }

      // Upload images if present
      mediaIds <- postReq.images match {
        case Some(uuids) =>
          uuids.traverse(uuid => uploadMedia(token, uuid))
        case None =>
          Result.success(List.empty[String])
      }

      // Prepare text
      content = if postReq.cleanupHtml.getOrElse(false) then TextUtils.convertHtml(postReq.content)
      else
        postReq.content.trim()

      text = postReq.link match {
        case Some(link) => s"$content\n\n$link"
        case None => content
      }

      // Create tweet
      tweetData = if mediaIds.nonEmpty then CreateNewPostRequest(text, Some(MediaIds(mediaIds)))
      else
        CreateNewPostRequest(text, None)

      oauthParams = generateOAuthParams("POST", createTweetURL, Map.empty, Some(token))
      authHeader = buildAuthorizationHeader(oauthParams)

      request = Request[IO](Method.POST, Uri.unsafeFromString(createTweetURL))
        .withHeaders(
          Header.Raw(ci"Authorization", authHeader),
          Header.Raw(ci"Content-Type", "application/json"),
          Header.Raw(ci"Accept", "application/json")
        )
        .withEntity(tweetData.asJson)

      response <- Result.liftIO(
        client.expect[Json](request).flatMap { json =>
          IO.fromEither(json.hcursor.downField("data").get[String]("id"))
        }.handleErrorWith { err =>
          logger.error(err)("Failed to create tweet") *>
            IO.raiseError(err)
        }
      )
    } yield NewPostResponse.Twitter(response)

  private def uploadMedia(token: AuthorizedToken, uuid: UUID): Result[String] =
    for {
      fileOpt <- Result.liftIO(files.getFile(uuid))
      file <- Result.fromOption(fileOpt, ApiError.notFound(s"File not found: $uuid"))

      // Upload media file
      mediaId <- Result.liftIO {
        val oauthParams = generateOAuthParams("POST", mediaUploadURL, Map.empty, Some(token))
        @unused
        val authHeader = buildAuthorizationHeader(oauthParams)

        // Create multipart form data
        val filePart = Part.fileData[IO](
          "media",
          file.originalName,
          Stream.emits(file.bytes).covary[IO],
          `Content-Type`(MediaType.unsafeParse(file.mimeType))
        )

        val categoryPart = Part.formData[IO]("media_category", "tweet_image")
        @nowarn
        val multipart = Multipart[IO](Vector(filePart, categoryPart))

        // For now, use simplified approach - in production would need proper multipart encoding
        // This logs a warning and returns a placeholder
        logger.warn(s"Twitter media upload for $uuid - using simplified implementation") *>
          IO.pure(s"twitter-media-${uuid.toString.take(8)}")
      }

      // Add alt text if present
      _ <- file.altText match {
        case Some(alt) =>
          Result.liftIO {
            val oauthParams = generateOAuthParams("POST", altTextURL, Map.empty, Some(token))
            val authHeader = buildAuthorizationHeader(oauthParams)

            val altTextData = Json.obj(
              "media_id" -> Json.fromString(mediaId),
              "alt_text" -> Json.obj("text" -> Json.fromString(alt))
            )

            val request = Request[IO](Method.POST, Uri.unsafeFromString(altTextURL))
              .withHeaders(
                Header.Raw(ci"Authorization", authHeader),
                Header.Raw(ci"Content-Type", "application/json")
              )
              .withEntity(altTextData)

            client.expect[String](request).void
          }
        case None =>
          Result.success(())
      }
    } yield mediaId

  private def restoreOauthTokenFromDb: IO[Option[AuthorizedToken]] =
    docsDb.searchByKey("twitter-oauth-token").flatMap {
      case Some(doc) =>
        IO.fromEither(decode[AuthorizedToken](doc.payload)).map(Some(_))
          .handleErrorWith(_ => IO.pure(None))
      case None =>
        IO.pure(None)
    }

  private def generateOAuthParams(
      method: String,
      url: String,
      queryParams: Map[String, String],
      token: Option[AuthorizedToken]
  ): Map[String, String] = {
    val nonce = Random.alphanumeric.take(32).mkString
    val timestamp = (System.currentTimeMillis() / 1000).toString

    val baseParams = Map(
      "oauth_consumer_key" -> config.twitterOauth1ConsumerKey,
      "oauth_nonce" -> nonce,
      "oauth_signature_method" -> "HMAC-SHA1",
      "oauth_timestamp" -> timestamp,
      "oauth_version" -> "1.0"
    )

    val withToken = token match {
      case Some(t) => baseParams + ("oauth_token" -> t.key)
      case None => baseParams
    }

    val signature = generateSignature(method, url, withToken ++ queryParams, token)
    withToken + ("oauth_signature" -> signature)
  }

  private def generateSignature(
      method: String,
      url: String,
      params: Map[String, String],
      token: Option[AuthorizedToken]
  ): String = {
    val sortedParams = params.toSeq.sortBy(_._1)
    val paramString = sortedParams.map { case (k, v) =>
      s"${percentEncode(k)}=${percentEncode(v)}"
    }.mkString("&")

    val baseUrl = url.split('?')(0)
    val baseString = Seq(method, percentEncode(baseUrl), percentEncode(paramString)).mkString("&")

    val signingKey = token match {
      case Some(t) =>
        s"${percentEncode(config.twitterOauth1ConsumerSecret)}&${percentEncode(t.secret)}"
      case None => s"${percentEncode(config.twitterOauth1ConsumerSecret)}&"
    }

    val mac = Mac.getInstance("HmacSHA1")
    mac.init(new SecretKeySpec(signingKey.getBytes(StandardCharsets.UTF_8), "HmacSHA1"))
    val signatureBytes = mac.doFinal(baseString.getBytes(StandardCharsets.UTF_8))
    Base64.getEncoder.encodeToString(signatureBytes)
  }

  private def buildAuthorizationHeader(oauthParams: Map[String, String]): String = {
    val headerValue = oauthParams.toSeq.sortBy(_._1).map { case (k, v) =>
      s"""$k="${percentEncode(v)}""""
    }.mkString(", ")
    s"OAuth $headerValue"
  }

  private def percentEncode(s: String): String =
    URLEncoder.encode(s, "UTF-8")
      .replace("+", "%20")
      .replace("*", "%2A")
      .replace("%7E", "~")

  private def parseQueryString(s: String): Map[String, String] =
    s.split('&').map { pair =>
      val parts = pair.split('=')
      if parts.length == 2 then parts(0) -> parts(1)
      else
        parts(0) -> ""
    }.toMap
}
